ios/brave-ios/Sources/BraveWallet/WalletUserAssetManager.swift

```diff
@@ -161,131 +173,240 @@ public class WalletUserAssetManager: WalletUserAssetManagerType, WalletObserverS
-        retrieveAllDataObserver().forEach { $0.userAssetUpdated() }
+        await refreshBalance(for: token)
+
+        // update WalletService
+        await walletService.removeUserAsset(token: token)
+        let success = await walletService.addUserAsset(token: token)
+
+        retrieveAllDataObserver().forEach {
+          $0.userAssetUpdated()
+        }
+
+        return success
+      } else {
+        // 1. auto-discovered is either visible or hidden asset not marked as deleted by user
+        // 2. custom asset is either visible or hidden and not marked as deleted by user
+        return false
       }
-    } else {  // asset does not exist in database
-      await WalletUserAsset.addUserAsset(
-        asset: asset
-      )
-      refreshBalance(for: asset)
-      retrieveAllDataObserver().forEach { $0.userAssetUpdated() }
+    } else {
+      // does not exist in database
+      // add it to database
+      await WalletUserAsset.addUserAsset(token: token)
+      await refreshBalance(for: token)
+
+      // update WalletService
+      let success: Bool
+      if !isAutoDiscovery {
+        success = await walletService.addUserAsset(token: token)
+      } else {
+        // auto-discovery assets already added to WalletService by core
+        success = true
+      }
+
+      retrieveAllDataObserver().forEach {
+        $0.userAssetUpdated()
+      }
+
+      return success
     }
   }
 
   @MainActor public func removeUserAsset(
-    _ asset: BraveWallet.BlockchainToken
-  ) async {
-    await WalletUserAsset.removeUserAsset(
-      asset: asset
+    _ token: BraveWallet.BlockchainToken
+  ) async -> Bool {
+    // update database
+    await updateUserAsset(
+      for: token,
+      visible: false,
+      isSpam: token.isSpam,
+      isDeletedByUser: true
     )
-    await removeBalances(for: asset)
-    retrieveAllDataObserver().forEach { $0.userAssetUpdated() }
+    await removeAssetBalances(for: token)
+
+    // remove user asset from WalletService
+    let success = await walletService.removeUserAsset(token: token)
+
+    retrieveAllDataObserver().forEach {
+      $0.userAssetUpdated()
+    }
+
+    return success
   }
 
   @MainActor public func updateUserAsset(
-    for asset: BraveWallet.BlockchainToken,
+    for token: BraveWallet.BlockchainToken,
     visible: Bool,
     isSpam: Bool,
     isDeletedByUser: Bool
   ) async {
     await WalletUserAsset.updateUserAsset(
-      for: asset,
+      for: token,
       visible: visible,
       isSpam: isSpam,
       isDeletedByUser: isDeletedByUser
     )
+    if token.visible != visible {
+      // update visibility in WalletService
+      let success = await walletService.setUserAssetVisible(
+        token: token,
+        visible: visible
+      )
+      if !success, !token.contractAddress.isEmpty {
+        // update the custom asset visibilty failed
+        // try to add this custom asset
+        await walletService.addUserAsset(token: token)
```

Comment:
If an asset is currently visible and user updates to not-visible, wouldn't we want to remove from wallet service not add it?

Reply:
Tested it and token will be stored with `visible: false`.